File: submission/source-code/activity5-junit/src/main/java/com/lab3/Calculator.java
package com.lab3;

/**
 * A robust calculator class providing basic arithmetic operations and utility functions.
 * Used for demonstrating white-box testing techniques with JUnit 5.
 */
public class Calculator {

    /**
     * Adds two integers.
     *
     * @param a first operand
     * @param b second operand
     * @return the sum of a and b
     */
    public int add(int a, int b) {
        return a + b;
    }

    /**
     * Subtracts the second integer from the first.
     *
     * @param a first operand
     * @param b second operand
     * @return the result of a - b
     */
    public int subtract(int a, int b) {
        return a - b;
    }

    /**
     * Multiplies two integers.
     *
     * @param a first operand
     * @param b second operand
     * @return the product of a and b
     */
    public int multiply(int a, int b) {
        return a * b;
    }

    /**
     * Divides the first integer by the second.
     *
     * @param a dividend
     * @param b divisor
     * @return the integer quotient of a / b
     * @throws ArithmeticException if b is zero
     */
    public int divide(int a, int b) {
        if (b == 0) {
            throw new ArithmeticException("Division by zero is not allowed");
        }
        return a / b;
    }

    /**
     * Computes base raised to the power of exp (base^exp).
     * Only supports non-negative exponents.
     *
     * @param base the base value
     * @param exp  the exponent (must be >= 0)
     * @return base raised to the power of exp
     * @throws IllegalArgumentException if exp is negative
     */
    public int power(int base, int exp) {
        if (exp < 0) {
            throw new IllegalArgumentException("Negative exponent not supported: " + exp);
        }
        int result = 1;
        for (int i = 0; i < exp; i++) {
            result *= base;
        }
        return result;
    }

    /**
     * Computes the modulo (remainder) of a divided by b.
     *
     * @param a dividend
     * @param b divisor
     * @return the remainder of a / b
     * @throws ArithmeticException if b is zero
     */
    public int modulo(int a, int b) {
        if (b == 0) {
            throw new ArithmeticException("Modulo by zero is not allowed");
        }
        return a % b;
    }

    /**
     * Checks if a number is positive (greater than zero).
     *
     * @param n the number to check
     * @return true if n > 0, false otherwise
     */
    public boolean isPositive(int n) {
        return n > 0;
    }

    /**
     * Returns the maximum of three integers.
     *
     * @param a first number
     * @param b second number
     * @param c third number
     * @return the largest value among a, b, and c
     */
    public int max(int a, int b, int c) {
        int max = a;
        if (b > max) {
            max = b;
        }
        if (c > max) {
            max = c;
        }
        return max;
    }

    /**
     * Converts a numeric score to a letter grade.
     * <ul>
     *   <li>A: 90-100</li>
     *   <li>B: 80-89</li>
     *   <li>C: 70-79</li>
     *   <li>D: 60-69</li>
     *   <li>F: 0-59</li>
     * </ul>
     *
     * @param score the numeric score (0-100)
     * @return the letter grade as a String
     * @throws IllegalArgumentException if score is outside the range 0-100
     */
    public String gradeScore(int score) {
        if (score < 0 || score > 100) {
            throw new IllegalArgumentException("Score must be between 0 and 100: " + score);
        }
        if (score >= 90) {
            return "A";
        } else if (score >= 80) {
            return "B";
        } else if (score >= 70) {
            return "C";
        } else if (score >= 60) {
            return "D";
        } else {
            return "F";
        }
    }
}

File: submission/source-code/activity4-mutation/calculator.py
"""
calculator.py — A Calculator class with arithmetic operations.

This module provides a Calculator class with methods that serve as the
subject for mutation testing. Each method has clear semantics and edge
cases to enable robust test design.
"""


class Calculator:
    """A simple calculator with arithmetic operations.

    All methods are static/instance methods that accept numeric inputs.
    """

    @staticmethod
    def add(a: float, b: float) -> float:
        """Return the sum of a and b.

        >>> Calculator.add(2, 3)
        5
        >>> Calculator.add(-1, 1)
        0
        """
        return a + b

    @staticmethod
    def subtract(a: float, b: float) -> float:
        """Return a minus b.

        >>> Calculator.subtract(5, 3)
        2
        >>> Calculator.subtract(3, 5)
        -2
        """
        return a - b

    @staticmethod
    def multiply(a: float, b: float) -> float:
        """Return the product of a and b.

        >>> Calculator.multiply(4, 3)
        12
        >>> Calculator.multiply(0, 5)
        0
        """
        return a * b

    @staticmethod
    def divide(a: float, b: float) -> float:
        """Return a divided by b.

        Args:
            a: Dividend.
            b: Divisor.

        Returns:
            Quotient as a float.

        Raises:
            ZeroDivisionError: If b is zero.

        >>> Calculator.divide(10, 2)
        5.0
        >>> Calculator.divide(7, 3)
        2.333...
        """
        if b == 0:
            raise ZeroDivisionError("Cannot divide by zero")
        return a / b

    @staticmethod
    def power(a: float, b: float) -> float:
        """Return a raised to the power b.

        >>> Calculator.power(2, 3)
        8
        >>> Calculator.power(5, 0)
        1
        >>> Calculator.power(4, 0.5)
        2.0
        """
        return a ** b

    @staticmethod
    def modulo(a: float, b: float) -> float:
        """Return a modulo b (remainder of a / b).

        Args:
            a: Dividend.
            b: Divisor.

        Returns:
            Remainder.

        Raises:
            ZeroDivisionError: If b is zero.

        >>> Calculator.modulo(10, 3)
        1
        >>> Calculator.modulo(10, 5)
        0
        """
        if b == 0:
            raise ZeroDivisionError("Cannot compute modulo with zero")
        return a % b
